### PR TITLE
Correctly fail jobs of any type 

### PR DESCRIPTION
**Pull request by sherifnada**
## What
Failed check/spec/discover jobs were not being transitioned to the `failed` state and stayed in `incomplete`. This in part exacerbated problems like #1462 .

This PR puts `JobStateTransitioner` (previously `JobRetrier`) in charge of this, and correctly transitions state. 

## How
1. Change `JobRetrier` to examine all incomplete jobs (regardless of type) and transitions them to failed if their attempt failed
2. Adds methods to list jobs with a set of statuses in `DefaultJobPersistence` and test cases. 

## Recommended reading order
1. `JobRetrier.java`
1. `DefaultJobPersistence.java` (and tests)
1. the rest

### CODE REVIEW

File: airbyte-scheduler/src/main/java/io/airbyte/scheduler/JobRetrier.java [hunk 1]
```diff
@@ -51,49 +51,53 @@ public JobRetrier(JobPersistence jobPersistence, Supplier<Instant> timeSupplier)
...
-      return persistence.listJobsWithStatus(JobConfig.ConfigType.SYNC, JobStatus.INCOMPLETE);
+      return persistence.listJobsWithStatus(JobStatus.INCOMPLETE);
     } catch (IOException e) {
       throw new RuntimeException("failed to fetch failed jobs", e);
     }
   }
 
   private boolean hasReachedMaxAttempt(Job job) {
-    return job.getAttemptsCount() >= MAX_ATTEMPTS;
+    if (job.getConfigType() == JobConfig.ConfigType.SYNC) {
```
Comment (michel-tricot) [inline airbyte-scheduler/src/main/java/io/airbyte/scheduler/JobRetrier.java:86]: you should use the io.airbyte.scheduler.Job#REPLICATION_TYPES instead.

File: airbyte-scheduler/src/main/java/io/airbyte/scheduler/JobRetrier.java [hunk 2]
```diff
@@ -51,49 +51,53 @@ public JobRetrier(JobPersistence jobPersistence, Supplier<Instant> timeSupplier)
...
+    } else {
+      return job.getAttemptsCount() >= 1;
+    }
   }
 
   private boolean shouldRetry(Job job) {
-    long lastRun = job.getUpdatedAtInSecond();
-    // todo (cgardens) - use exponential backoff.
-    return lastRun < timeSupplier.get().getEpochSecond() - TimeUnit.MINUTES.toSeconds(RETRY_WAIT_MINUTES);
+    if (job.getConfigType() == JobConfig.ConfigType.SYNC) {
```
Comment (michel-tricot) [inline airbyte-scheduler/src/main/java/io/airbyte/scheduler/JobRetrier.java:94]: Same

File: airbyte-scheduler/src/main/java/io/airbyte/scheduler/persistence/DefaultJobPersistence.java [hunk 3]
```diff
@@ -278,14 +281,26 @@ private Job getJob(DSLContext ctx, long jobId) {
...
 
   @Override
-  public List<Job> listJobsWithStatus(ConfigType configType, JobStatus status) throws IOException {
+  public List<Job> listJobsWithStatus(JobStatus status) throws IOException {
+    return listJobsWithStatus(Sets.newHashSet(ConfigType.values()), status);
+  }
+
+  @Override
+  public List<Job> listJobsWithStatus(Set<ConfigType> configTypes, JobStatus status) throws IOException {
+    String[] clauses = configTypes.stream().map(type -> "CAST(config_type AS VARCHAR) = ? ").toArray(String[]::new);
```
Comment (michel-tricot) [inline airbyte-scheduler/src/main/java/io/airbyte/scheduler/persistence/DefaultJobPersistence.java:290]: Same here. there are example in this file of how to query for the replication jbos

**Review by michel-tricot — APPROVED** [review]

File: airbyte-scheduler/src/main/java/io/airbyte/scheduler/JobRetrier.java (hunk 1 above)
Comment (sherifnada) [reply airbyte-scheduler/src/main/java/io/airbyte/scheduler/JobRetrier.java]: done for all

File: airbyte-scheduler/src/main/java/io/airbyte/scheduler/JobRetrier.java [hunk 4]
```diff
@@ -51,49 +50,54 @@ public JobRetrier(JobPersistence jobPersistence, Supplier<Instant> timeSupplier)
...
-      return persistence.listJobsWithStatus(JobConfig.ConfigType.SYNC, JobStatus.INCOMPLETE);
+      return persistence.listJobsWithStatus(JobStatus.INCOMPLETE);
     } catch (IOException e) {
       throw new RuntimeException("failed to fetch failed jobs", e);
     }
   }
 
   private boolean hasReachedMaxAttempt(Job job) {
-    return job.getAttemptsCount() >= MAX_ATTEMPTS;
+
```
Comment (michel-tricot) [inline airbyte-scheduler/src/main/java/io/airbyte/scheduler/JobRetrier.java:85]: NIT: new line

Comment (sherifnada) [inline airbyte-scheduler/src/main/java/io/airbyte/scheduler/JobRetrier.java:85]:
```suggestion
```

**Comment by cgardens** [comment]
@sherifnada should we also make sure we do not block spec, check, and discover jobs on each other? it seems like if we add some random suffix in the scope we could get closer to the our desired behavior.

**Pull request merged**